import csv

STAR_OUTDIR='bam/{source}/RNA/STAR/'

ss=[]

with open(config['samplesheet'],'r') as csvfile:
    reader = csv.DictReader(csvfile,delimiter='\t')
    for row in reader:
        ss.append(row)


sources = set()
for s in ss:
    sources.add(s['source'])


def source2fastq(wc):
    ret={'fastq':[],
         'rg':[]}
    for row in filter(lambda res:res['source']==wc.source,ss):
        ret['fastq']+=[row['read1'],row['read2']]
        ret['rg']+=[{'ID':row['result_id'],
                     'LB':row['library_id'],
                     'SM':row['source']}]
    return(ret)


localrules: final

rule final:
        params: batch="-q ccr -l nodes=1"
        input: expand(STAR_OUTDIR+"/Aligned.sortedByCoord.out.bam",source=sources)
        
rule rnaalignment:
    input:  lambda wildcards: source2fastq(wildcards)['fastq']
    output: STAR_OUTDIR+"/Aligned.sortedByCoord.out.bam",
            STAR_OUTDIR+"/Chimeric.out.junction",
            STAR_OUTDIR+"/Chimeric.out.sam"
    params:  batch="-q ccr -l nodes=1:g128",outdir=STAR_OUTDIR
    priority: 10
    threads: 32
    version: "2.4.1a"
    run:
        read1 = ','.join(input[0::2])
        read2 = ','.join(input[1::2])
        rg = " , ".join(["ID:{ID} LB:{LB} SM:{SM} PL:Illumina".format(**s) for s in source2fastq(wildcards)['rg']])
        #rg = ",".join(["ID:"+res[0] for res in source2fastq(wildcards)['rg']])
        shell("""mkdir -p {params.outdir}
clearscratch
module load STAR/{version}
module load samtools
STAR --genomeDir /data/CCRBioinfo/public/STAR/hg19_gencode14_ov60/ \
        --outFileNamePrefix {params.outdir}/ \
        --readFilesIn {read1} {read2} \
        --runThreadN {threads} \
        --outSAMstrandField intronMotif \
        --outFilterIntronMotifs RemoveNoncanonicalUnannotated \
        --outReadsUnmapped None --chimSegmentMin 15 \
        --chimJunctionOverhangMin 15 \
        --alignMatesGapMax 200000 \
        --alignIntronMax 200000 \
        --outSAMtype BAM SortedByCoordinate \
        --outTmpDir /scratch/tmp \
        --outSAMattrRGline {rg} \
        --outSAMunmapped Within \
        --sjdbOverhang 60
""")

rule rnapostalignment:
    input: bam="bam/{source}/RNA/STAR/Aligned.out.bam",chimeric="bam/{source}/RNA/STAR/Chimeric.out.sam"
    output: bam="bam/{source}/RNA/STAR/{source}.bam",chimeric=temp("bam/{source}/RNA/STAR/chimeric.bam")
    params: batch='-l nodes=1:gpfs -q ccr',outbam='bam/{source}/RNA/STAR/chimeric'
    threads: 2
    shell: """samtools view -bS {input.chimeric} | samtools sort - {params.outbam} \
java -jar /usr/local/picard/MergeSamFiles.jar I={output.chimeric} \
   I={input.bam} O={output.bam} AS=true"""
